streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java

```diff
@@ -835,69 +839,87 @@ public synchronized final TopologyBuilder addInternalTopic(String topicName) {
-                sourceTopics.addAll(((SourceNodeFactory) nodeFactory).topics);
+                sourceNodes.add((SourceNodeFactory) nodeFactory);
             } else if (nodeFactory instanceof ProcessorNodeFactory) {
-                sourceTopics.addAll(findSourceTopicsForProcessorParents(((ProcessorNodeFactory) nodeFactory).parents));
+                sourceNodes.addAll(findSourcesForProcessorParents(((ProcessorNodeFactory) nodeFactory).parents));
             }
         }
-        return sourceTopics;
+        return sourceNodes;
     }
 
-    private void connectStateStoreNameToSourceTopics(final String stateStoreName,
-                                                     final ProcessorNodeFactory processorNodeFactory) {
+    private void connectStateStoreNameToSourceTopicsOrPattern(final String stateStoreName,
+                                                              final ProcessorNodeFactory processorNodeFactory) {
+
+        final Set<String> sourceTopics =  new HashSet<>();
+        final Set<Pattern> sourcePatterns = new HashSet<>();
+        final Set<SourceNodeFactory> sourceNodesForParent = findSourcesForProcessorParents(processorNodeFactory.parents);
+
+        for (SourceNodeFactory sourceNodeFactory : sourceNodesForParent) {
+            if (sourceNodeFactory.pattern != null) {
+                sourcePatterns.add(sourceNodeFactory.pattern);
+            } else {
+                sourceTopics.addAll(sourceNodeFactory.topics);
+            }
+        }
+
+
+        if (sourceTopics.isEmpty() && sourcePatterns.isEmpty()) {
+            throw new TopologyBuilderException("can't find source topic or source Pattern for state store " +
+                    stateStoreName);
+        }
 
         // we should never update the mapping from state store names to source topics if the store name already exists
         // in the map; this scenario is possible, for example, that a state store underlying a source KTable is
         // connecting to a join operator whose source topic is not the original KTable's source topic but an internal repartition topic.
-        if (stateStoreNameToSourceTopics.containsKey(stateStoreName)) {
-            return;
+
+        if (!stateStoreNameToSourceTopics.containsKey(stateStoreName) && !sourceTopics.isEmpty()) {
+            stateStoreNameToSourceTopics.put(stateStoreName,
+                    Collections.unmodifiableSet(sourceTopics));
         }
 
-        final Set<String> sourceTopics = findSourceTopicsForProcessorParents(processorNodeFactory.parents);
-        if (sourceTopics.isEmpty()) {
-            throw new TopologyBuilderException("can't find source topic for state store " +
-                    stateStoreName);
+        if (!stateStoreNameToSourceRegex.containsKey(stateStoreName) && !sourcePatterns.isEmpty()) {
+            stateStoreNameToSourceRegex.put(stateStoreName,
+                    Collections.unmodifiableSet(sourcePatterns));
+
```

Comment:
I think they are two conditions:

1. `sourceTopics.isEmpty() && sourcePatterns.isEmpty()` means we cannot find any source node with the processor node tracing back. I think this should never happen?

2. `stateStoreNameToSourceTopics.containsKey(stateStoreName) XOR stateStoreNameToSourceRegex.containsKey(stateStoreName)` means that this state store has already been connected to SOME source nodes already (think: connect store to node1, then connect store to node2, where node1 and node2 trace back to two different topics). In this case we need to update the map by adding the sourceTopics / Partitions to the value instead of overwriting it?